File: Algorithms/Sorting/bucket_sort.cpp
// Name : Anjali Kumari
// Email : LIT2016033@iiitl.ac.in

// This program contains implementation of Bucket Sort Algorithm.

#include<iostream>
#include<algorithm>
#include<vector>
#include<utility>
using namespace std;

#define bucket_size 10

void bucketSort(float arr[], long n)
{  
    vector<float> bucket[bucket_size];
    //we make empty bucket vectors of size bucket_size , vector is choosen instead of
	//2-d array because we can dynamically regulate size of vector.
	
    for(long i=0;i<n;i++)
    {
       long d = arr[i]*n; //Index determination in a bucket.
       bucket[d].push_back(arr[i]);
    }
    for(long i=0;i<bucket_size;i++)
    {
       	sort(bucket[i].begin(),bucket[i].end());
    }
    long j=0,index=0;
    while(j!=bucket_size)
    {
       for(long i=0;i<bucket[j].size();i++)
       {
            arr[index++]=bucket[j][i];
       }
       j++;
    }
}
 
//Function to print the elements of Array
void printArray(float arr[],long arr_length)
{
    for(long i=0; i<arr_length; i++)
    {
      printf("%f ", arr[i]);
    }
    printf("\n");
}

// Driver program to test bucketSort function
int main()
{
    long n;
    printf("Please enter the number of elements in Array."); 
    scanf("%lld",&n);
    float arr[n];
    printf("\nPlease enter the elements of array to be sorted.\n");
    
    for(long i=0; i<n ;i++)
    {
      scanf("%f",&arr[i]);
    }
    
    printf("The entered elements are :\n");
    printArray(arr,n);
    
    //function to sort elements in ascending order using Bucket sort
    bucketSort(arr, n); 
    printf("The sorted array is: \n");
    printArray(arr, n);
    return 0;
}

//Bucket sort assumes that the input is drawn from a uniform distribution.
//Sample Input : 6 0.2324 0.9876 0.2345 0.4456 0.7878 0.8672
//Output : 0.2324 0.2345 0.4456 0.7878 0.8672 0.9876 
//Space Complexity : O(n+k) 
//Average Case and Best Case Time Complexity : O(n+k)
//Worst Case Time Complexity : O(n^2) 
//n is the number of elements to be sorted and k is the number of buckets
//Bucket Sort is a sorting method that subdivides the given data into various buckets .
//Bucket sort’s best case occurs when the data being sorted can be distributed between 
//the buckets perfectly. If the values are sparsely allocated over the possible value range, 
//a larger bucket size is better since the buckets will likely be more evenly distributed.
//Bucket sort performs at its worst when all elements are allocated to the same bucket.
//I have used inbuilt sort in for all buckets.It has time complexity of O(n log n).
//Insertion sort can also be used for sorting these buckets ,but if single bucket will 
//contain all elements then Insertion Sort will be a bad choice .

File: Algorithms/Sorting/Java/BucketSort.java
//Bucket Sort in java

import java.util.*;


public class BucketSort {


        //Driver method
	public static void main(String[] args) {
		
               Scanner sc=new Scanner(System.in);

               //Enter the size of array.
               int n=sc.nextInt();

               double a[]=new double[n];

               //Entering array elements.
               for(int i=0;i<n;i++)
                  a[i]=sc.nextDouble();

	       double res[] = new double[a.length];
		
               //Calling a function named buket.
	       res = bucket(a,a.length);
                  

	       //Printing the sorted array. 
               for(int i=0;i<res.length;i++)
	         System.out.println(res[i]);
	}

	//Method implementing bucket sort.
	public static double[] bucket(double arr[], int n)
	{
	    //  Create n empty buckets
	    ArrayList<ArrayList<Double>> b=new ArrayList<ArrayList<Double>>();
	    
	    for(int i=0;i<n;i++)
	    	b.add(new ArrayList<Double>());
	    	    
	    
	    //  Put array elements in different buckets
	    for (int i=0; i<n; i++)
	    {
	       // val is storing the index for an element in bucket
               int val = (int) (n*arr[i]);

	       b.get(val).add(arr[i]);
	    }
	 
	    //  Sort all buckets individualy.
	    for (int i=0; i<n; i++)
	       Collections.sort(b.get(i));
	 
	    //  Concatenate all buckets into arr[]
	    int num = 0;
	    for (int i = 0; i < n; i++)
	        for (int j = 0; j < b.get(i).size(); j++)
	          arr[num++] = b.get(i).get(j);
	
	   return arr;
	
	
	}

}




//Time Complexity - The average case complexity is O(n) 
//but in worst case the complexity is O(n^2) since in this case all the elements are stored in only one bucket.
